File: 2236-maximum-twin-sum-of-a-linked-list/maximum-twin-sum-of-a-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def pairSum(self, head: Optional[ListNode]) -> int:
        slow = head
        fast = head
        prev = None
        while fast and fast.next:
            fast = fast.next.next
            next_node = slow.next
            slow.next = prev
            prev, slow = slow, next_node
        
        res = 0
        while slow:
            res = max(res, slow.val + prev.val)
            slow = slow.next
            prev = prev.next
        return res

            

File: 2236-maximum-twin-sum-of-a-linked-list/maximum-twin-sum-of-a-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    int pairSum(ListNode* head) {
        ListNode* slow = head, *fast = head;
        stack<int> st;
        while (fast && fast->next) {
            slow = slow->next;
            fast = fast->next->next;
        }

        while (slow) {
            st.push(slow->val);
            slow = slow->next;
        }

        int res = 0;
        slow = head;
        while (!st.empty()) {
            res = max(res, st.top() + slow->val);
            slow = slow->next;
            st.pop();
        }

        return res;
    }
};